lib/spike.rb

```diff
@@ -0,0 +1,257 @@
+require 'forwardable'
+require 'json'
+require 'pathname'
+require 'active_support/inflector'
+require 'oas_parser'
+
+module Spike
+
+  class Endpoint
+    class PositionalParameterizer
+      def parameterize(args)
+        "#{args.join(", ")}, options = {}"
+      end
+    end
+
+    class KwargsParameterizer
+      def parameterize(args)
+        "#{args.map {|arg| arg + ":"}.join(", ")}, **options"
+      end
+    end
+
+    extend Forwardable
+
+    VERB_PRIORITY = %w(GET POST)
+
+    delegate [:documentation_url] => :definition
+
+    attr_reader :definition, :parameterizer
+    def initialize(oas_endpoint, parameterizer: Spike::Endpoint::PositionalParameterizer)
+      @definition    = oas_endpoint
+      @parameterizer = parameterizer.new
+    end
+
+    def to_s
+      [
+        tomdoc,
+        method_definition,
+        alias_definition,
+      ].compact.join("\n")
+    end
+
+    def tomdoc
+      <<-TOMDOC.chomp
+      # #{definition.summary}
+      #
+      # #{parameter_documentation.join("\n      # ")}
+      # @return #{return_type_description} #{return_value_description}
+      # @see #{definition.raw["externalDocs"]["url"]}
+      TOMDOC
+    end
+
+    def method_definition
+      <<-DEF.chomp
+      def #{method_name}(#{parameters})
+        #{subresource?? subresource_method_implementation : method_implementation}
+      end
+      DEF
+    end
+
+    def alias_definition
+      return unless alternate_name
+      "      alias :#{alternate_name} :#{method_name}"
+    end
+
+    def singular?
+      definition.path.path.include? "_id"
+    end
+
+    def subresource?
+      namespace.split("_").count > 1
+    end
+
+    def method_implementation
+      [
+        *option_overrides,
+        api_call,
+      ].reject(&:empty?).join("\n        ")
+    end
+
+    # TODO. possible breaking change as the url param isn't a required param 
```

Comment:
Previously, the `status` methods (i.e.  for `get-a-single-deployment-status`, `list-deployment-statuses`) had a required [`deployment_url` parameter](https://github.com/octokit/octokit.rb/blob/a1eeda95fa5fcbb532c411e3851e2d8ab894a5ea/lib/octokit/client/deployments.rb#L51). However, the [current OpenAPI spec](https://github.com/octokit/routes/blob/master/openapi/api.github.com/operations/repos/list-deployment-statuses.json#L12) doesn't have this as a required parameter and has `repo` and `deployment_id` instead. 

Going forward, do we want to focus on supporting the existing format and ignore the new parameters so it's not a breaking change?

Reply:
Making it a breaking change 🎉